src/gui/inputdevices/InputUtils.h

```diff
@@ -1,5 +1,3 @@
-#pragma once
```

Comment:
**NEVER NEVER NEVER EVER remove the include barrier in a header file!!!**

I can't stress that point enough.

If you have to do that, this means you have circular dependencies in your headers. You may fix them by predeclaring structs or classes in this header. If you just remove the include barrier you might end up with duplicate definitions of classes or structs with really obscure compiler warnings. Believe me that is something you do not want.

Reply:
This was exactly the pointer I was missing. Being not that familiar with the cpp toolchain yet, I wasn't shure while the linking was throwing an error with `undefined reference`. I guess this was a result of a circular dependency then.

It felt weird that suddenly withouth the `#pragma` it worked that's why I asked explicitly for feedback that. Thanks. And thank you even more for the general advice :pray: .

I guees I'll find a better place then for this util function :detective:

Reply:
The C compiler goes through your source code (.c file) and includes headers mentioned with `#include`. This operation is literally pasting the content of the file at the line of the macro. C itself has the restriction that all structs, classes and signatures, used for an invocation, must be declared priorly. This is because the compiler stores each struct or signature in a table to match jumping addresses. The actual definition may then follow at some later point. In your case this means you can just predeclare the necessary structs, classes and functions within the header. 

https://www.geeksforgeeks.org/what-are-forward-declarations-in-c/

Reply:
One question in general: Would you consider it good practice to have a `#pragma once` at the top of every header file, or should it only be used in specific scenarios.

Reply:
Every header should have a `#pragma once` or be encapsulated in a

```
#ifdef __HEADER_XYZ__
#define __HEADER_XYZ__

...

#endif
```

It is crucial that headers are only ever included once per source file. As generally header files reference other header files this is not always the case. Hence what you do is to tell the preprocessor to only include a file once even if it is referenced multiple times.